Update go.opentelemetry.io dependencies to v1.33.0/v0.58.0

dashpole opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Prerequisite for https://github.com/kubernetes/kubernetes/pull/127053.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
```

dashpole commented:
cc @kubernetes/sig-instrumentation-leads

dashpole commented:
/retest

dashpole commented:
cc @dims

This does get rid of one unwanted dependency.

dims commented:
@dashpole thanks! looks like we bring in `github.com/planetscale/vtprotobuf` as indirect, do we know from where?

![image](https://c.tenor.com/ebYFr6Kj_80AAAAd/tenor.gif)

dims commented:
@dashpole thanks! I am ok with it being an indirect dependency for now. LGTM in spirit.

cc @liggitt just in case he spots something, we still have ways to go before we can land anything in master

liggitt commented on hack/unwanted-dependencies.json:
```diff
@@ -192,7 +192,6 @@
         "github.com/prometheus/client_golang",
         "go.etcd.io/etcd/client/v2",
         "go.etcd.io/etcd/server/v3",
-        "go.opentelemetry.io/contrib/instrumentation/github.com/emicklei/go-restful/otelrestful",
```
🎉

liggitt commented:
> It comes through: google.golang.org/grpc -> github.com/envoyproxy/go-control-plane -> github.com/planetscale/vtprotobuf

ironic that google.golang.org/grpc pulls in an alternate proto library 🤦 

I'm _really_ surprised google.golang.org/grpc is pulling in github.com/envoyproxy/go-control-plane ... looks like that's not an easy link to cut 😢 

I think we should add that to our unwanted dependencies list to avoid picking up more links accidentally or starting to actually vendor it without knowing in the future... I'm really antsy about getting *more* proto implementations in tree, especially if they may not play nice around a common required version of core proto libraries / types

liggitt commented on vendor/github.com/emicklei/go-restful/v3/jsr311.go:
```diff
@@ -155,7 +155,7 @@ func (r RouterJSR311) detectRoute(routes []Route, httpRequest *http.Request) (*R
 		if (method == http.MethodPost ||
 			method == http.MethodPut ||
-			method == http.MethodPatch) && length == "" {
+			method == http.MethodPatch) && (length == "" || length == "0") {
```
uh... this looks incorrect at first glance... is an explicitly empty (zero-length) body invalid?

things like https://lists.w3.org/Archives/Public/ietf-http-wg/2010JulSep/0276.html led me to think a zero-length body on put/post/patch was valid

liggitt replied:
commented at https://github.com/emicklei/go-restful/pull/543#issuecomment-2495000742

dashpole commented:
/retest

fedebongio commented:
/remove-sig api-machinery

dgrisonnet commented:
/triage accepted
/assign @dims

dims commented:
@dashpole @dgrisonnet this will look a lot better after a rebase :)

dashpole commented:
@dims rebased.  I also updated it to the next version (from v1.32.0 -> v1.33.0).  That brought in one additional transitive dependency from otel, `go.opentelemetry.io/auto/sdk`, which is part of the otel api library.

dims commented:
thanks @dashpole we talked about `github.com/planetscale/vtprotobuf` earlier so it's not a surprise. thanks for calling out `go.opentelemetry.io/auto/sdk`.

/approve
/lgtm

dims commented:
/assign @liggitt @saschagrunert

dashpole commented:
/test pull-kubernetes-e2e-capz-windows-master

dashpole commented:
@dims rebased

dims commented:
/approve
/lgtm

poke  @liggitt @saschagrunert !!

dashpole commented:
@dims @liggitt I've rebased and reverted the updates to github.com/emicklei/go-restful and go.opentelemetry.io/contrib/instrumentation/github.com/emicklei/go-restful/otelrestful, because of the issue you found.  Hopefully that is enough to unblock this :).

dims commented:
@dashpole a few transitive dependencies are making me queasy a bit (but they are transitive!). Let's run it by Jordan.

```
❯ go mod graph | grep github.com/planetscale/vtprotobuf
google.golang.org/grpc@v1.69.4 github.com/planetscale/vtprotobuf@v0.6.1-0.20240319094008-0393e58bdf10
❯ go mod graph | grep github.com/GoogleCloudPlatform/opentelemetry-operations-go/detectors/gcp
google.golang.org/grpc@v1.69.4 github.com/GoogleCloudPlatform/opentelemetry-operations-go/detectors/gcp@v1.24.2
❯ go mod graph | grep go.opentelemetry.io/contrib/detectors/gcp
google.golang.org/grpc@v1.69.4 go.opentelemetry.io/contrib/detectors/gcp@v1.31.0
```

other than that the auto/sdk is a new dependency, we just have to tolerate that update, so good there.

thanks,
dims

dashpole commented:
I'm the owner of both github.com/GoogleCloudPlatform/opentelemetry-operations-go/detectors/gcp and go.opentelemetry.io/contrib/detectors/gcp FWIW.  They are minimal libraries for fetching resource attributes from the GKE metadata server using http calls.  Since we don't use the grpc-provided stats handler library for OTel, we aren't using the library.

I'm surprised to see those being pulled in... I think grpc's OpenTelemetry stats handler should be in its own module to avoid this, so I opened https://github.com/grpc/grpc-go/issues/8092.  I'll add these to unwanted dependencies.

We discussed github.com/planetscale/vtprotobuf earlier.  I forgot to add it to unwanted dependencies again.

dashpole commented:
Done in https://github.com/kubernetes/kubernetes/pull/128919/commits/f9445994a9f47a02e4baa26f5dba797a828ff1fa

dims commented:
thanks @dashpole !

dashpole commented:
Looks like the new dependencies from grpc are intentional (from https://github.com/grpc/grpc-go/issues/8092#issuecomment-2662085755):

> This is the PR that merged the openetelemetry module into the main grpc-go module: [#7759](https://github.com/grpc/grpc-go/pull/7759)
> 
> There is a discussion about a similar concern of introducing a gcp specific dependency: [#7759 (comment)](https://github.com/grpc/grpc-go/pull/7759#discussion_r1819903090)
> 
> Quoting from the discussion:
> 
> > The decision was made to release these CSM Observability/OpenTelemetry API's as stable. Rather than have OpenTelemetry as it's own module, we decided to release these features as simply part of our top level go module, as C-Core does. So I think this will be left here. Note that this is simply listing the dependencies. It does not affect binary size or performance unless the feature that uses the dependency is enabled (in this case, CSM Observability).
> 
> > Note that we also put in some safeguards to help protect this from accidentally getting imported into any of the main grpc packages in [#7766](https://github.com/grpc/grpc-go/pull/7766). There are some nasty ways they could sneak in by moving common functionality to an internal package and then adding an internal dependency on that from somewhere else. We have similar concerns about other things affecting binary size and dependency trees like xds.

liggitt commented:
That's really unfortunate... even if they don't get linked in, the dependencies will make the graph more complex and impact go's selected versions of dependencies :-/

dashpole commented:
Ok, I've downgraded back to v1.33.0/v0.58.0, which doesn't bring in the new GCP dependencies (via grpc).  This unblocks my tracing work in 1.33 while we discuss on the gRPC issue.  I've also rebased again

dims commented:
> Ok, I've downgraded back to v1.33.0/v0.58.0, which doesn't bring in the new GCP dependencies (via grpc). This unblocks my tracing work in 1.33 while we discuss on the gRPC issue.

Thanks @dashpole !!

cc @liggitt @saschagrunert (for remaining approvals)

liggitt commented:
/approve
/hold cancel

needs rebase on https://github.com/kubernetes/kubernetes/pull/130474, sorry

dashpole committed:
update go.opentelemetry.io dependencies to v1.33.0/v0.58.0

dashpole commented:
/test pull-kubernetes-e2e-gce

dims commented:
/approve
/lgtm
/skip
